src/app/workflow/dag/wdl-viewer/wdl-viewer.component.ts

```diff
@@ -64,29 +67,42 @@ export class WdlViewerComponent implements AfterViewInit, OnDestroy {
           return;
         }
 
-        if (files && files.length > 0) {
+        // Add delay so parent component can finish rendering it's view before the status is set
+        this.wdlViewerQuery.selectEntity(this.version.id).pipe(take(1), delay(0))
+          .subscribe((cached: WdlViewerPipeline) => {
+            this.wdlViewerError = false;
 
-          // Create the Epam WDL visualization and attach the result to the DOM element. Stop subscribing after first completion
-          this.wdlViewerService.create(files, this.workflow, this.version).pipe(take(1), finalize(() => {
-            this.loading = false;
-            this.versionChanged = false;
-          }))
-            .subscribe((res: WdlViewerPipeline) => {
-              this.visualizer.attachTo(res.model[0]);
-              this.wdlViewerError = false;
+            if (cached) {
+
+              // Attach cached pipeline to the DOM element
+              this.visualizer.attachTo(cached.model[0]);
+              this.clearProgressBar();
               this.wdlViewerService.setStatus(true);
-            },
-              (error) => {
-                this.errorMessage = error || 'Unknown Error';
-                this.wdlViewerError = true;
-                this.diagram.nativeElement.remove();
-                this.wdlViewerService.setStatus(false);
-              });
-        } else {
-          this.wdlViewerService.setStatus(false);
-          this.wdlViewerError = true;
-          this.loading = false;
-        }
+
+            } else if (files && files.length > 0) {
+
+              // Create the Epam WDL visualization and attach the result to the DOM element. Stop subscribing after first completion
+              this.wdlViewerService.create(files, this.workflow, this.version).pipe(take(1), finalize(() => {
+                this.clearProgressBar();
+              }))
+                .subscribe((res: WdlViewerPipeline) => {
+                    res.id = this.version.id;
+                    this.visualizer.attachTo(res.model[0]);
+                    this.wdlViewerService.update(this.workflow.id, res);
```

Comment:
I haven't tried running this, but won't this cause the selectEntity subscription on line 71 to get an observable and you'll end up attaching to visualizer twice?

Reply:
This doesn't attach the object to the visualizer twice because the selectEntity subscription on line 71 pipes the `take` operator. The operator limits the subscription to emit only 1 value (or entity in this case) and then stops subscribing. On the other hand, if the subscription piped `takeUntil(this.ngUnsubscribe)` then the visualizer would attach it twice

I also verified that this code doesn't attach twice with breakpoints in Chrome debugger.

Reply:
Thanks for verifying!

I'm still wondering if it wouldn't be simpler though to just drive the visualization through the store. Right now you have attachTo in two places. If you just always visualized from the store, instead of sometimes visualizing from the store and sometimes from the result of an Ajax call.

Reply:
That would be possible if I utilize `takeUntil(this.ngUnsubscribe)` instead of `take(1)` on line 71, and remove `attachTo` from line 90. This code snippet produces the situation described in the original comment where the selectEntity subscription emits at two times: first time, the cached object is null; second time, the cached object exists.

Should this be the way to go?

Reply:
I think that's the preferred way, yes. Look at [the diagram](https://netbasal.gitbook.io/akita/); in a perfect world the component is getting data via the query.